lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,519 @@
+    for (var element in this) {
+      yield* expend(index++, element);
+    }
+  }
+
+  /// The first element satisfying [test], or `null` if there are none.
+  T /*?*/ firstWhereOrNull(bool test(T element)) {
+    for (var element in this) {
+      if (test(element)) return element;
+    }
+    return null;
+  }
+
+  /// The first element, or `null` if the iterable is empty.
+  T /*?*/ get firstOrNull {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) return iterator.current;
+    return null;
+  }
+
+  /// The last element satisfying [test], or `null` if there are none.
+  T /*?*/ lastWhereOrNull(bool test(T element)) {
+    T /*?*/ result;
+    for (var element in this) {
+      if (test(element)) result = element;
+    }
+    return result;
+  }
+
+  /// The last element, or `null` if the iterable is empty.
+  T /*?*/ get lastOrNull {
+    if (isEmpty) return null;
+    return last;
+  }
+
+  /// The single element satisfying [test].
+  ///
+  /// Returns `null` if there are either no elements
+  /// or more than one element satisfying [test].
+  T /*?*/ singleWhereOrNull(bool test(T element)) {
+    T /*?*/ result;
+    var found = false;
+    for (var element in this) {
+      if (test(element)) {
+        if (!found) {
+          result = element;
+          found = true;
+        } else {
+          return null;
+        }
+      }
+    }
+    return result;
+  }
+
+  /// The single element of the iterable, or `null`.
+  ///
+  /// The value is `null` if the iterable is empty
+  /// or it contains more than one element.
+  T /*?*/ get singleOrNull {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) {
+      var result = iterator.current;
+      if (!iterator.moveNext()) {
+        return result;
+      }
+    }
+    return null;
+  }
+
+  /// Groups elements by [keyOf] then folds the elements in each group.
+  ///
+  /// A key is found for each element using [keyOf].
+  /// Then the elements with the same key are all folded using [combine].
+  /// The first call to [combine] for a particular key receives [null] as
+  /// the previous value, the remaining ones receive the result of the previous call.
+  ///
+  /// Can be used to _group_ elements into arbitrary collections.
+  /// For example [groupSetsBy] could be written as:
+  /// ```dart
+  /// iterable.groupFoldBy(keyOf,
+  ///     (Set<T>? previous, T element) => (previous ?? <T>{})..add(element));
+  /// ````
+  Map<K, G> groupFoldBy<K, G>(
+      K keyOf(T element), G combine(G /*?*/ previous, T element)) {
+    var result = <K, G>{};
+    for (var element in this) {
+      var key = keyOf(element);
+      result[key] = combine(result[key], element);
+    }
+    return result;
+  }
+
+  /// Groups elements into sets by [keyOf].
+  Map<K, Set<T>> groupSetsBy<K>(K keyOf(T element)) {
+    var result = <K, Set<T>>{};
+    for (var element in this) {
+      var key = keyOf(element);
+      result[key] = (result[key] ?? {})..add(element);
```

Comment:
I'm not sure if it actually makes a difference (or if we should be trying to micro-optimize without benchmarks) but I _think_ this is faster on dart2js with `??=`

```suggestion
      (result[key] ??= {}).add(element);
```